File: assist/python/HmTask/down_load_pic/download_pic.py
import pandas as pd
import os
import requests
from queue import Queue
from threading import Thread
from concurrent.futures import ThreadPoolExecutor, as_completed
from PIL import Image, ImageDraw, ImageFont
from abc import ABC, abstractmethod
from pathlib import Path
import logging
import configparser

# 配置日志
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

# 读取配置文件
config = configparser.ConfigParser()
config.read('config.ini')

# 从配置文件读取配置项
download_dir_name = config.get('directories', 'download_dir')
final_dir_name = config.get('directories', 'final_dir')
max_download_workers = config.getint('threading', 'max_download_workers')
max_conversion_workers = config.getint('threading', 'max_conversion_workers')

# 定义下载函数
def download_file(url, local_path):
    try:
        response = requests.get(url)
        response.raise_for_status()  # 检查请求是否成功
        with open(local_path, 'wb') as file:
            file.write(response.content)
        logging.info(f"下载完成: {local_path}")
        return local_path
    except Exception as e:
        logging.error(f"下载失败: {url} -> {local_path}, 错误: {e}")
        return None

# 定义转换和添加文字的函数
def convert_and_add_text(src_path, final_path):
    try:
        # 读取图片
        img = Image.open(src_path).convert('RGBA')

        # 获取图片尺寸
        width, height = img.size

        # 提取文件名中的文本
        filename = os.path.basename(final_path)
        text = filename.split('_')[0].split('-')[1]

        # 设置字体和文字高度
        font_path = r"C:\Windows\Fonts\STKAITI.TTF"  # 华文楷体字体路径
        font = ImageFont.truetype(font_path, 30)

        # 计算文字的位置
        text_bbox = font.getbbox(text)
        text_width = text_bbox[2] - text_bbox[0]
        text_height = text_bbox[3] - text_bbox[1]

        # 确保文字宽度不超过图片宽度
        if text_width > width:
            font = ImageFont.truetype(font_path, int(20 * (width / text_width)))

        # 计算文字的位置
        text_bbox = font.getbbox(text)
        text_width = text_bbox[2] - text_bbox[0]
        text_height = text_bbox[3] - text_bbox[1]
        text_x = (width - text_width) // 2
        text_y = height - text_height - 15  # 文字距离底部的距离

        # 创建带有透明背景的新图像
        text_img = Image.new('RGBA', (text_width, text_height), (0, 0, 0, 0))
        draw = ImageDraw.Draw(text_img)

        # 添加文字
        draw.text((0, 0), text, fill=(255, 255, 255, 255), font=font)

        # 将带有文字的透明图像与原始图像合并
        img.paste(text_img, (text_x, text_y), text_img)

        # 保存最终图片
        img = img.convert('RGB')
        img.save(final_path, format='JPEG')
        # os.remove(src_path)
        logging.info(f"转换并添加文字完成: {final_path}")
    except Exception as e:
        logging.error(f"转换并添加文字失败: {src_path} -> {final_path}, 错误: {e}")

# 定义基类
class BaseWorkerThread(Thread, ABC):
    def __init__(self, task_queue, max_workers):
        super().__init__()
        self.task_queue = task_queue
        self.max_workers = max_workers

    def run(self):
        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
            self.execute_tasks(executor)

    @abstractmethod
    def execute_tasks(self, executor):
        pass

# 定义生产者基类
class ProducerThreadBase(BaseWorkerThread, ABC):
    def __init__(self, task_queue, max_workers, items):
        super().__init__(task_queue, max_workers)
        self.items = items

    def execute_tasks(self, executor):
        for item in self.items:
            self.execute_task(item, executor)

    @abstractmethod
    def execute_task(self, item, executor):
        pass

# 定义消费者基类
class ConsumerThreadBase(BaseWorkerThread, ABC):
    def execute_tasks(self, executor):
        while True:
            try:
                item = self.task_queue.get(timeout=1)
                self.execute_task(item, executor)
                self.task_queue.task_done()
            except Empty:
                break

    @abstractmethod
    def execute_task(self, item, executor):
        pass

# 定义生产者线程类
class CustomProducerThread(ProducerThreadBase):
    def __init__(self, urls, names, max_download_workers, download_dir, final_dir, task_queue):
        items = list(zip(urls, names))
        super().__init__(task_queue, max_download_workers, items)
        self.download_dir = download_dir
        self.final_dir = final_dir

    def execute_task(self, item, executor):
        url, name = item
        file_name = name + ".jpg"
        cur_path = os.path.join(self.download_dir, file_name)
        final_path = os.path.join(self.final_dir, file_name)

        if os.path.exists(final_path):
            logging.info(f"最终文件已存在: {final_path}")
            return

        if os.path.exists(cur_path):
            # 如果临时文件存在，则直接提交转换任务
            self.task_queue.put((cur_path, final_path))
        else:
            # 否则提交下载任务，并添加回调函数
            future = executor.submit(download_file, url, cur_path)
            future.add_done_callback(lambda f, q=self.task_queue, p=final_path: q.put((f.result(), p)) if f.result() is not None else None)

# 定义消费者线程类
class CustomConsumerThread(ConsumerThreadBase):
    def execute_task(self, item, executor):
        src_path, final_path = item
        executor.submit(convert_and_add_text, src_path, final_path)

# 设置线程池的最大线程数
def process_images(file_path, sheet_name):
    # 文件路径和目录设置
    cur_dir = os.path.dirname(file_path)
    download_dir = os.path.join(cur_dir, download_dir_name)
    if not os.path.exists(download_dir):
        os.makedirs(download_dir)
    final_dir = os.path.join(cur_dir, final_dir_name)
    if not os.path.exists(final_dir):
        os.makedirs(final_dir)

    # 读取Excel文件
    df = pd.read_excel(file_path, sheet_name=sheet_name)

    # 获取URL和名称列
    urls = df['src'].tolist()
    names = df['dest_name'].tolist()

    # 创建任务队列
    task_queue = Queue()

    # 创建生产者线程
    producer_thread = CustomProducerThread(urls, names, max_download_workers, download_dir, final_dir, task_queue)
    producer_thread.start()

    # 创建消费者线程
    consumer_thread = CustomConsumerThread(task_queue, max_conversion_workers)
    consumer_thread.start()

    # 等待生产者线程完成
    producer_thread.join()

    # 等待消费者线程完成
    consumer_thread.join()

    # 确保所有任务完成
    task_queue.join()

    logging.info("所有任务已完成")

# 主程序入口
if __name__ == "__main__":
    # 文件路径和目录设置
    file_path = r"F:\教程\多肉\322种常见多肉植物，1分钟认全.xlsx"
    sheet_name = "多肉"  # 假设工作表的名字为“多肉”

    process_images(file_path, sheet_name)

File: assist/python/HmTask/down_load_pic/download_pic.cpp
#include <iostream>
#include <string>
#include <vector>
#include <thread>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <filesystem>
#include <curl/curl.h> // libcurl
#include <cimg/CImg.h> // CImg library for image processing
#include <spdlog/spdlog.h> // spdlog for logging
#include <inih/ini.h> // inih for reading INI files

// 日志初始化
void init_logging() {
    auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
    auto logger = std::make_shared<spdlog::logger>("app_logger", console_sink);
    spdlog::register_logger(logger);
    spdlog::set_level(spdlog::level::info);
}

// 从配置文件读取配置项
struct Config {
    std::string download_dir;
    std::string final_dir;
    int max_download_workers;
    int max_conversion_workers;
};

Config read_config(const std::string& config_path) {
    Config config;
    ini_parse(config_path.c_str(), [](void* user, const char* section, const char* name, const char* value) {
        Config* cfg = static_cast<Config*>(user);
        if (!strcmp(section, "directories")) {
            if (!strcmp(name, "download_dir")) {
                cfg->download_dir = value;
            } else if (!strcmp(name, "final_dir")) {
                cfg->final_dir = value;
            }
        } else if (!strcmp(section, "threading")) {
            if (!strcmp(name, "max_download_workers")) {
                cfg->max_download_workers = std::stoi(value);
            } else if (!strcmp(name, "max_conversion_workers")) {
                cfg->max_conversion_workers = std::stoi(value);
            }
        }
    }, &config);
    return config;
}

size_t write_data(void* contents, size_t size, size_t nmemb, void* userp) {
    ((std::string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}

// 下载文件
bool download_file(const std::string& url, const std::string& local_path) {
    CURL* curl;
    CURLcode res;
    std::string readBuffer;

    curl_global_init(CURL_GLOBAL_DEFAULT);
    curl = curl_easy_init();
    if (curl) {
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
        res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);

        if (res == CURLE_OK) {
            std::ofstream file(local_path);
            if (file.is_open()) {
                file << readBuffer;
                file.close();
                spdlog::info("下载完成: {}", local_path);
                return true;
            }
        }
    }
    spdlog::error("下载失败: {} -> {}, 错误: {}", url, local_path, curl_easy_strerror(res));
    return false;
}

// 转换并添加文字
void convert_and_add_text(const std::string& src_path, const std::string& final_path) {
    CImg<unsigned char> image(src_path.c_str());
    if (image.is_empty()) {
        spdlog::error("无法打开图片: {}", src_path);
        return;
    }

    // 这里省略了具体的文字添加逻辑，可以参考Python版本实现细节
    // ...

    image.save(final_path.c_str());
    spdlog::info("转换并添加文字完成: {}", final_path);
}

// 任务队列
class TaskQueue {
public:
    void push(Task task) {
        std::lock_guard<std::mutex> lock(m_mutex);
        m_queue.push(std::move(task));
        m_condition.notify_one();
    }

    Task pop() {
        std::unique_lock<std::mutex> lock(m_mutex);
        m_condition.wait(lock, [this] { return !m_queue.empty(); });
        Task task = std::move(m_queue.front());
        m_queue.pop();
        return task;
    }

private:
    std::queue<Task> m_queue;
    std::mutex m_mutex;
    std::condition_variable m_condition;
};

// 生产者线程
void producer_thread(const std::vector<std::string>& urls, const std::vector<std::string>& names, TaskQueue& task_queue, const Config& config) {
    for (size_t i = 0; i < urls.size(); ++i) {
        std::string file_name = names[i] + ".jpg";
        std::string cur_path = config.download_dir + "/" + file_name;
        std::string final_path = config.final_dir + "/" + file_name;

        if (std::filesystem::exists(final_path)) {
            spdlog::info("最终文件已存在: {}", final_path);
            continue;
        }

        if (std::filesystem::exists(cur_path)) {
            task_queue.push([cur_path, final_path]() { convert_and_add_text(cur_path, final_path); });
        } else {
            if (download_file(urls[i], cur_path)) {
                task_queue.push([cur_path, final_path]() { convert_and_add_text(cur_path, final_path); });
            }
        }
    }
}

// 消费者线程
void consumer_thread(TaskQueue& task_queue) {
    while (true) {
        auto task = task_queue.pop();
        task();
    }
}

int main() {
    init_logging();
    Config config = read_config("config.ini");

    // 创建必要的目录
    std::filesystem::create_directories(config.download_dir);
    std::filesystem::create_directories(config.final_dir);

    // 示例数据
    std::vector<std::string> urls = {"http://example.com/image1.jpg", "http://example.com/image2.jpg"};
    std::vector<std::string> names = {"image1", "image2"};

    TaskQueue task_queue;
    std::thread producer(producer_thread, urls, names, std::ref(task_queue), config);
    std::thread consumer(consumer_thread, std::ref(task_queue));

    producer.join();
    consumer.join();

    spdlog::info("所有任务已完成");
    return 0;
}